taichi/program/ndarray.h

```diff
@@ -31,13 +31,21 @@ class Ndarray {
   intptr_t get_device_allocation_ptr_as_int() const;
   std::size_t get_element_size() const;
   std::size_t get_nelement() const;
+  ~Ndarray();
 
  private:
   DeviceAllocation ndarray_alloc_{kDeviceNullAllocation};
+  // Invariant:
+  //   data_ptr_ is not nullptr iff arch is a llvm backend
   uint64_t *data_ptr_{nullptr};
   uint64_t *device_allocation_ptr_{nullptr};
   std::size_t nelement_{1};
   std::size_t element_size_{1};
+  // Ndarrays manage their own |DeviceAllocation| so this must be shared with
+  // |OpenGlRuntime|. Without the ownership, when the program exits |device_|
+  // mighted be destructed earlier than Ndarray object, leaving a segfault when
```

Comment:
```suggestion
  // might be destructed earlier than Ndarray object, leaving a segfault when
```

Reply:
Hmm, I see why this is necessary, but I wonder if there's a better way to handle this. The other way around is to track all the Ndarrays in the Program, which is more complicated to implement. Does torch track its allocated tensors, though?

Reply:
As we discussed offline, I'll keep the dealloc logic in ndarray destructor for now, until we start to implement DTR :D